Tested on cluster.
The build keeps failing on ```CDAPOperationalStatsTest ``` and fails locally too

```
java.lang.AssertionError: expected:<0> but was:<1>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:743)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:555)
	at org.junit.Assert.assertEquals(Assert.java:542)
	at co.cask.cdap.operations.cdap.CDAPOperationalStatsTest.test(CDAPOperationalStatsTest.java:94)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```
From the test  case it looks like the expected value should be 1 but on release/4.1 branch it passes all the time with expected value 0. I am not sure what is the issue here and how it passes on release

--- a/cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java
+++ b/cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java
@@ -220,7 +220,8 @@ ApplicationLifecycleService extends AbstractIdleService {
       ArtifactId artifactId = appSpec.getArtifactId();
       ArtifactSummary artifactSummary = artifactId == null ?
         new ArtifactSummary(appSpec.getName(), null) : ArtifactSummary.from(artifactId);
-      ApplicationRecord record = new ApplicationRecord(artifactSummary, appId, appSpec.getDescription());
+      ApplicationRecord record = new ApplicationRecord(artifactSummary, appId, appSpec.getDescription(),
+                                                       ownerAdmin.getImpersonationPrincipal(appId));
       if (predicate.apply(record)) {
         appRecords.add(record);
       }

--- a/cdap-cli-tests/src/test/java/co/cask/cdap/cli/CLIMainTest.java
+++ b/cdap-cli-tests/src/test/java/co/cask/cdap/cli/CLIMainTest.java
@@ -251,6 +251,8 @@ public class CLIMainTest extends CLITestBase {
     testCommandOutputContains(cli, "list streams", streamId);
     testCommandOutputContains(cli, "list dataset instances", datasetId);
     testCommandOutputContains(cli, "delete app " + ConfigTestApp.NAME, "Successfully");
+    testCommandOutputContains(cli, "delete dataset instance " + datasetId, "Successfully deleted");
+    testCommandOutputContains(cli, "delete stream " + streamId, "Successfully deleted stream");
   }
 
   @Test
@@ -270,6 +272,43 @@ public class CLIMainTest extends CLITestBase {
     testCommandOutputContains(cli, "list streams", streamId);
     testCommandOutputContains(cli, "list dataset instances", datasetId);
     testCommandOutputContains(cli, "delete app " + ConfigTestApp.NAME, "Successfully");
+    testCommandOutputContains(cli, "delete dataset instance " + datasetId, "Successfully deleted");
+    testCommandOutputContains(cli, "delete stream " + streamId, "Successfully deleted stream");
+  }
+
+  @Test
+  public void testAppOwner() throws Exception {
+    // load an artifact
+    File appJarFile = createAppJarFile(ConfigTestApp.class);
+    String streamId = ConfigTestApp.DEFAULT_STREAM;
+    String datasetId = ConfigTestApp.DEFAULT_TABLE;
+    testCommandOutputContains(cli, String.format("load artifact %s name %s version %s", appJarFile.getAbsolutePath(),
+                                                 "OwnedConfigTestAppArtifact", V1),
+                              "Successfully added artifact");
+
+    // create a config file which has principal information
+    File configFile = new File(TMP_FOLDER.newFolder(), "testOwnerConfigFile.txt");
+    try (BufferedWriter writer = Files.newWriter(configFile, Charsets.UTF_8)) {
+      writer.write(String.format("{\n\"%s\":\"%s\"\n}", ArgumentName.PRINCIPAL, "user/host.net@kdc.net"));
+      writer.close();
+    }
+
+    // create app with the config containing the principla
+    testCommandOutputContains(cli, String.format("create app %s %s %s %s %s",
+                                                 "OwnedApp", "OwnedConfigTestAppArtifact", V1,
+                                                 ArtifactScope.USER, configFile.getAbsolutePath()),
+                              "Successfully created application");
+
+    /// ensure that the app details have owner information
+    testCommandOutputContains(cli, "list apps", "user/host.net@kdc.net");
+
+    // clean up
+    if (!appJarFile.delete()) {
+      LOG.warn("Failed to delete temporary app jar file: {}", appJarFile.getAbsolutePath());
+    }
+    testCommandOutputContains(cli, "delete app " + "OwnedApp", "Successfully");
+    testCommandOutputContains(cli, "delete dataset instance " + datasetId, "Successfully deleted");
+    testCommandOutputContains(cli, "delete stream " + streamId, "Successfully deleted stream");
   }
 
   @Test

--- a/cdap-cli/src/main/java/co/cask/cdap/cli/command/app/CreateAppCommand.java
+++ b/cdap-cli/src/main/java/co/cask/cdap/cli/command/app/CreateAppCommand.java
@@ -65,16 +65,18 @@ public class CreateAppCommand extends AbstractAuthCommand {
     ArtifactSummary artifact = new ArtifactSummary(artifactName, artifactVersion, artifactScope);
 
     JsonObject config = new JsonObject();
+    String ownerPrincipal = null;
     String configPath = arguments.getOptional(ArgumentName.APP_CONFIG_FILE.toString());
     if (configPath != null) {
       File configFile = resolver.resolvePathToFile(configPath);
       try (FileReader reader = new FileReader(configFile)) {
         AppRequest<JsonObject> appRequest = GSON.fromJson(reader, configType);
         config = appRequest.getConfig();
+        ownerPrincipal = appRequest.getOwnerPrincipal();
       }
     }
 
-    AppRequest<JsonObject> appRequest = new AppRequest<>(artifact, config);
+    AppRequest<JsonObject> appRequest = new AppRequest<>(artifact, config, ownerPrincipal);
     applicationClient.deploy(appId, appRequest);
     output.println("Successfully created application");
   }
@@ -93,7 +95,8 @@ public class CreateAppCommand extends AbstractAuthCommand {
                            "configuration is needed, it must be given as a file whose contents are a JSON object " +
                            "containing the application config. For example, the file contents could contain: " +
                            "'{ \"config\": { \"stream\": \"purchases\" } }'. In this case, the application would " +
-                           "receive '{ \"stream\": \"purchases\" }' as its config object.",
+                           "receive '{ \"stream\": \"purchases\" }' as its config object. Finally, an optional " +
+                           "principal may be given",
       Fragment.of(Article.A, ElementType.APP.getName()), ApplicationId.DEFAULT_VERSION);
   }
 }

--- a/cdap-cli/src/main/java/co/cask/cdap/cli/command/app/ListAppsCommand.java
+++ b/cdap-cli/src/main/java/co/cask/cdap/cli/command/app/ListAppsCommand.java
@@ -58,13 +58,14 @@ public class ListAppsCommand extends AbstractAuthCommand {
       }
     }
     Table table = Table.builder()
-      .setHeader("id", "appVersion", "description", "artifactName", "artifactVersion", "artifactScope")
+      .setHeader("id", "appVersion", "description", "artifactName", "artifactVersion", "artifactScope", "principal")
       .setRows(appClient.list(cliConfig.getCurrentNamespace(), artifactNames, artifactVersion),
         new RowMaker<ApplicationRecord>() {
           @Override
           public List<?> makeRow(ApplicationRecord object) {
             return Lists.newArrayList(object.getName(), object.getAppVersion(), object.getDescription(),
-              object.getArtifact().getName(), object.getArtifact().getVersion(), object.getArtifact().getScope());
+              object.getArtifact().getName(), object.getArtifact().getVersion(), object.getArtifact().getScope(),
+              object.getOwnerPrincipal());
           }
         }).build();
     cliConfig.getTableRenderer().render(cliConfig, output, table);

--- a/cdap-client/src/main/java/co/cask/cdap/client/ApplicationClient.java
+++ b/cdap-client/src/main/java/co/cask/cdap/client/ApplicationClient.java
@@ -123,8 +123,8 @@ public class ApplicationClient {
    * specified artifact name and version.
    *
    * @param namespace the namespace to list applications from
-   * @param artifactName the name of the artifact to filter by. If null, no filtering will be done
-   * @param artifactVersion the version of the artifact to filter by. If null, no filtering will be done
+   * @param artifactName the name of the artifact to filter by. If null, no filtering will be done.
+   * @param artifactVersion the version of the artifact to filter by. If null, no filtering will be done.
    * @return list of {@link ApplicationRecord ApplicationRecords}.
    * @throws IOException if a network error occurred
    * @throws UnauthenticatedException if the request is not authorized successfully in the gateway server
@@ -142,8 +142,8 @@ public class ApplicationClient {
    * specified artifact name and version.
    *
    * @param namespace the namespace to list applications from
-   * @param artifactName the name of the artifact to filter by. If null, no filtering will be done
-   * @param artifactVersion the version of the artifact to filter by. If null, no filtering will be done
+   * @param artifactName the name of the artifact to filter by. If null, no filtering will be done.
+   * @param artifactVersion the version of the artifact to filter by. If null, no filtering will be done.
    * @return list of {@link ApplicationRecord ApplicationRecords}.
    * @throws IOException if a network error occurred
    * @throws UnauthenticatedException if the request is not authorized successfully in the gateway server
@@ -163,8 +163,8 @@ public class ApplicationClient {
    * the specified artifact names and the specified artifact version.
    *
    * @param namespace the namespace to list applications from
-   * @param artifactNames the set of artifact names to allow. If empty, no filtering will be done
-   * @param artifactVersion the version of the artifact to filter by. If null, no filtering will be done
+   * @param artifactNames the set of artifact names to allow. If empty, no filtering will be done.
+   * @param artifactVersion the version of the artifact to filter by. If null, no filtering will be done.
    * @return list of {@link ApplicationRecord ApplicationRecords}.
    * @throws IOException if a network error occurred
    * @throws UnauthenticatedException if the request is not authorized successfully in the gateway server
@@ -182,8 +182,8 @@ public class ApplicationClient {
    * the specified artifact names and the specified artifact version.
    *
    * @param namespace the namespace to list applications from
-   * @param artifactNames the set of artifact names to allow. If empty, no filtering will be done
-   * @param artifactVersion the version of the artifact to filter by. If null, no filtering will be done
+   * @param artifactNames the set of artifact names to allow. If empty, no filtering will be done.
+   * @param artifactVersion the version of the artifact to filter by. If null, no filtering will be done.
    * @return list of {@link ApplicationRecord ApplicationRecords}.
    * @throws IOException if a network error occurred
    * @throws UnauthenticatedException if the request is not authorized successfully in the gateway server
@@ -475,7 +475,7 @@ public class ApplicationClient {
   /**
    * Deploys an application.
    *
-   * @param jarFile jar file of the application to deploy
+   * @param jarFile JAR file of the application to deploy
    * @throws IOException if a network error occurred
    * @deprecated since 4.0.0. Use {@link #deploy(NamespaceId, File)} instead.
    */
@@ -487,7 +487,7 @@ public class ApplicationClient {
   /**
    * Deploys an application.
    *
-   * @param jarFile jar file of the application to deploy
+   * @param jarFile JAR file of the application to deploy
    * @throws IOException if a network error occurred
    */
   public void deploy(NamespaceId namespace, File jarFile) throws IOException, UnauthenticatedException {
@@ -499,7 +499,7 @@ public class ApplicationClient {
    * Deploys an application with a serialized application configuration string.
    *
    * @param namespace namespace to which the application should be deployed
-   * @param jarFile jar file of the application to deploy
+   * @param jarFile JAR file of the application to deploy
    * @param appConfig serialized application configuration
    * @throws IOException if a network error occurred
    * @throws UnauthenticatedException if the request is not authorized successfully in the gateway server
@@ -515,17 +515,36 @@ public class ApplicationClient {
    * Deploys an application with a serialized application configuration string.
    *
    * @param namespace namespace to which the application should be deployed
-   * @param jarFile jar file of the application to deploy
+   * @param jarFile JAR file of the application to deploy
    * @param appConfig serialized application configuration
    * @throws IOException if a network error occurred
    * @throws UnauthenticatedException if the request is not authorized successfully in the gateway server
    */
   public void deploy(NamespaceId namespace, File jarFile,
                      @Nullable String appConfig) throws IOException, UnauthenticatedException {
+    deploy(namespace, jarFile, appConfig, null);
+  }
+
+  /**
+   * Deploys an application with a serialized application configuration string and an owner.
+   *
+   * @param namespace namespace to which the application should be deployed
+   * @param jarFile JAR file of the application to deploy
+   * @param appConfig serialized application configuration
+   * @param ownerPrincipal the principal of application owner
+   * @throws IOException if a network error occurred
+   * @throws UnauthenticatedException if the request is not authorized successfully in the gateway server
+   */
+  public void deploy(NamespaceId namespace, File jarFile,
+                     @Nullable String appConfig,
+                     @Nullable String ownerPrincipal) throws IOException, UnauthenticatedException {
     Map<String, String> headers = new HashMap<>();
     headers.put("X-Archive-Name", jarFile.getName());
     if (appConfig != null) {
       headers.put("X-App-Config", appConfig);
+    }
+    if (ownerPrincipal != null) {
+      headers.put("X-Principal", ownerPrincipal);
     }
     deployApp(namespace, jarFile, headers);
   }
@@ -534,7 +553,7 @@ public class ApplicationClient {
    * Deploys an application with an application configuration object.
    *
    * @param namespace namespace to which the application should be deployed
-   * @param jarFile jar file of the application to deploy
+   * @param jarFile JAR file of the application to deploy
    * @param appConfig application configuration object
    * @throws IOException if a network error occurred
    * @throws UnauthenticatedException if the request is not authorized successfully in the gateway server
@@ -550,7 +569,7 @@ public class ApplicationClient {
    * Deploys an application with an application configuration object.
    *
    * @param namespace namespace to which the application should be deployed
-   * @param jarFile jar file of the application to deploy
+   * @param jarFile JAR file of the application to deploy
    * @param appConfig application configuration object
    * @throws IOException if a network error occurred
    * @throws UnauthenticatedException if the request is not authorized successfully in the gateway server
@@ -651,7 +670,7 @@ public class ApplicationClient {
   }
 
   /**
-   * Lists all programs of some type.
+   * Lists all programs of a type.
    *
    * @param programType type of the programs to list
    * @return list of {@link ProgramRecord}s
@@ -666,7 +685,7 @@ public class ApplicationClient {
   }
 
   /**
-   * Lists all programs of some type.
+   * Lists all programs of a type.
    *
    * @param programType type of the programs to list
    * @return list of {@link ProgramRecord}s
@@ -724,7 +743,7 @@ public class ApplicationClient {
   }
 
   /**
-   * Lists programs of some type belonging to an application.
+   * Lists programs of a type belonging to an application.
    *
    * @param app the application
    * @param programType type of the programs to list
@@ -741,7 +760,7 @@ public class ApplicationClient {
   }
 
   /**
-   * Lists programs of some type belonging to an application.
+   * Lists programs of a type belonging to an application.
    *
    * @param app the application
    * @param programType type of the programs to list
@@ -765,7 +784,7 @@ public class ApplicationClient {
   }
 
   /**
-   * Lists programs of some type belonging to an application.
+   * Lists programs of a type belonging to an application.
    *
    * @param app the application
    * @return Map of {@link ProgramType} to list of {@link ProgramRecord}s
@@ -781,7 +800,7 @@ public class ApplicationClient {
   }
 
   /**
-   * Lists programs of some type belonging to an application.
+   * Lists programs of a type belonging to an application.
    *
    * @param app the application
    * @return Map of {@link ProgramType} to list of {@link ProgramRecord}s
@@ -803,7 +822,7 @@ public class ApplicationClient {
   }
 
   /**
-   * Lists programs belonging to an application.
+   * Lists all programs belonging to an application.
    *
    * @param app the application
    * @return List of all {@link ProgramRecord}s
@@ -819,7 +838,7 @@ public class ApplicationClient {
   }
 
   /**
-   * Lists programs belonging to an application.
+   * Lists all programs belonging to an application.
    *
    * @param app the application
    * @return List of all {@link ProgramRecord}s

--- a/cdap-operational-stats-core/src/test/java/co/cask/cdap/operations/cdap/CDAPOperationalStatsTest.java
+++ b/cdap-operational-stats-core/src/test/java/co/cask/cdap/operations/cdap/CDAPOperationalStatsTest.java
@@ -91,7 +91,10 @@ public class CDAPOperationalStatsTest {
     Assert.assertEquals(1, transactions.getNumInProgressTransactions());
     Assert.assertEquals(1, transactions.getNumInvalidTransactions());
     Assert.assertEquals(1, transactions.getNumCommittingChangeSets());
-    Assert.assertEquals(0, transactions.getNumCommittedChangeSets());
+    // the entites.collect calls appLifecycleService.getApps which fetches the owner principal of the deployed
+    // app making the committed transaction 1 here. We assert for 1 or greater because we might add other such calls
+    // which will cause a transaction to be completed.
+    Assert.assertTrue(transactions.getNumCommittedChangeSets() >= 1);
     Assert.assertTrue(transactions.getVisibilityUpperBound() > TEST_START_TIME);
     Assert.assertTrue(transactions.getSnapshotTime() > TEST_START_TIME);
     Assert.assertTrue(transactions.getReadPointer() > TEST_START_TIME);

--- a/cdap-proto/src/main/java/co/cask/cdap/proto/ApplicationRecord.java
+++ b/cdap-proto/src/main/java/co/cask/cdap/proto/ApplicationRecord.java
@@ -18,8 +18,10 @@ package co.cask.cdap.proto;
 
 import co.cask.cdap.proto.artifact.ArtifactSummary;
 import co.cask.cdap.proto.id.ApplicationId;
+import com.google.gson.annotations.SerializedName;
 
 import java.util.Objects;
+import javax.annotation.Nullable;
 
 /**
  * Represents item in the list from /apps
@@ -31,15 +33,23 @@ public class ApplicationRecord {
   private final String version;
   private final String description;
   private final ArtifactSummary artifact;
+  @SerializedName("principal")
+  private final String ownerPrincipal;
 
 
   public ApplicationRecord(ArtifactSummary artifact, ApplicationId appId, String description) {
+    this(artifact, appId, description, null);
+  }
+
+  public ApplicationRecord(ArtifactSummary artifact, ApplicationId appId, String description,
+                           @Nullable String ownerPrincipal) {
     this.type = "App";
     this.artifact = artifact;
     this.name = appId.getApplication();
     this.description = description;
     this.version = appId.getVersion();
     this.id = appId.getApplication();
+    this.ownerPrincipal = ownerPrincipal;
   }
 
   public ArtifactSummary getArtifact() {
@@ -67,6 +77,11 @@ public class ApplicationRecord {
     return description;
   }
 
+  @Nullable
+  public String getOwnerPrincipal() {
+    return ownerPrincipal;
+  }
+
   @Override
   public boolean equals(Object o) {
     if (this == o) {
@@ -82,12 +97,13 @@ public class ApplicationRecord {
       Objects.equals(name, that.name) &&
       Objects.equals(version, that.version) &&
       Objects.equals(description, that.description) &&
-      Objects.equals(artifact, that.artifact);
+      Objects.equals(artifact, that.artifact) &&
+      Objects.equals(ownerPrincipal, that.ownerPrincipal);
   }
 
   @Override
   public int hashCode() {
-    return Objects.hash(type, name, version, description, artifact);
+    return Objects.hash(type, name, version, description, artifact, ownerPrincipal);
   }
 
   @Override
@@ -98,6 +114,7 @@ public class ApplicationRecord {
       ", version='" + version + '\'' +
       ", description='" + description + '\'' +
       ", artifact=" + artifact +
+      ", ownerPrincipal=" + ownerPrincipal +
       '}';
   }
 }

